app/models/api/openid_connect/authorization.rb

```diff
@@ -0,0 +1,79 @@
+module Api
+  module OpenidConnect
+    class Authorization < ActiveRecord::Base
+      belongs_to :user
+      belongs_to :o_auth_application
+
+      validates :user, presence: true
+      validates :o_auth_application, presence: true
+      validates :user, uniqueness: {scope: :o_auth_application}
+      validate :validate_scope_names
+      serialize :scopes, JSON
+
+      has_many :o_auth_access_tokens, dependent: :destroy
+      has_many :id_tokens, dependent: :destroy
+
+      before_validation :setup, on: :create
+
+      scope :with_redirect_uri, ->(given_uri) { where redirect_uri: given_uri }
+
+      SCOPES = %w(openid sub aud name nickname profile picture read write)
+
+      def setup
+        self.refresh_token = SecureRandom.hex(32)
+      end
+
+      def validate_scope_names
+        return unless scopes
+        scopes.each do |scope|
+          errors.add(:scope, "is not a valid scope name") unless SCOPES.include? scope
+        end
+      end
+
+      def accessible?(required_scopes=nil)
+        Array(required_scopes).all? { |required_scope|
+          scopes.include? required_scope
+        }
+      end
+
+      def create_code
+        SecureRandom.hex(32).tap do |code|
+          self.code = code
+          save
```

Comment:
`update(code: code)` or perhaps `update!(code: code)` so we notice if it goes wrong for some reason.